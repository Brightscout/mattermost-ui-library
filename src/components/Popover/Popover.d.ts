import React from 'react';
import {OverlayTriggerProps} from 'react-bootstrap';

/**
 * Types for position
 */
type placementType =
    | 'auto-start'
    | 'auto'
    | 'auto-end'
    | 'top-start'
    | 'top'
    | 'top-end'
    | 'right-start'
    | 'right'
    | 'right-end'
    | 'bottom-end'
    | 'bottom'
    | 'bottom-start'
    | 'left-end'
    | 'left'
    | 'left-start';

export interface PopoverProps extends Pick<OverlayTriggerProps, 'children'> {

    /**
     * Placement of the tooltip with respect to the wrapped component
     *
     * @default `auto`
     */
    placement?: placementType;

    /**
     * Component to be rendered in the body of the popover
     */
    popoverBody: React.ReactNode;

    /**
     * Elevation for the popover
     *
     * @default false
     */
    elevation?: boolean;

    /**
     * To override or extend the styles applied to the component
     */
    className?: string;
}
