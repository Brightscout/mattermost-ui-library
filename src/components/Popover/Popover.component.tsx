import React, {useEffect, useRef, useState} from 'react';
import {OverlayTrigger} from 'react-bootstrap';
import {OverlayTriggerRenderProps} from 'react-bootstrap/esm/OverlayTrigger';

import {PopoverProps} from './Popover';
import {EventPassContainer, StyledPopover} from './Popover.styles';

/**
 * Popover component
 *
 * @example Correct usage
 *
 * ```ts
 * <Popover
 *  placement='auto'
 *  popoverBody={<Component />}
 * >
 *  <Icon name='menu dots' />
 * </Popover>
 * ```
 */
export const Popover = (props: PopoverProps) => {
    const {
        children,
        popoverBody,
        elevation,
        placement = 'auto',
        className = '',
    } = props;
    const [show, setShow] = useState(false);

    const EventPassContainerRef = useRef<HTMLDivElement>(null);

    /**
	 * Function which closes the popover on clicking anywhere outside the popover wrapper
	 * @param {MouseEvent} e - gets the element on which the onClick event is acted upon
	 */
    const onPopoverCloseHandler = (e: MouseEvent) =>
        e.target !== EventPassContainerRef.current && setShow(false);

    // Function to render the popover body
    const renderedPopover = (renderedPopoverProps: OverlayTriggerRenderProps) => (
        <StyledPopover
            elevation={elevation}
            className={`mm-popover ${className}`}
            {...renderedPopoverProps}
        >
            {popoverBody}
        </StyledPopover>
    );

    // On clicking anywhere other than the wrapped component the pop-over closes
    useEffect(() => {
        document.body.addEventListener('click', onPopoverCloseHandler);

        return () => {
            document.body.removeEventListener('click', onPopoverCloseHandler);
        };
    }, []);

    return (
        <OverlayTrigger
            key={placement}
            placement={placement}
            overlay={renderedPopover}
            trigger={'click'}
            show={show}
        >
            {/*
                Since all child components won't be spreading props, the event won't be able to reach down.
                Therefore passing the event to a div which wraps the children
            */}
            <EventPassContainer
                ref={EventPassContainerRef}
                onClick={() => {
                    setShow((prev) => !prev);
                }}
            >
                {children}
            </EventPassContainer>
        </OverlayTrigger>
    );
};
