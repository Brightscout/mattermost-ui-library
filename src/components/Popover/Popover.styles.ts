import styled from 'styled-components';
import {Popover} from 'react-bootstrap';

import Colors from '@Styles/colorsForJs.module.scss';

// Styles for the pop over
export const StyledPopover = styled(Popover)<{ elevation: boolean }>(
    ({elevation}) => ({
        margin: 8,
        borderRadius: 4,
        boxShadow: elevation ? Colors.elevation1 : 'none',

        // removing arrow
        '& .arrow': {
            display: 'none',
        },

        '& *': {
            cursor: 'pointer',
        },
    }),
);

// Styles for the event pass container
export const EventPassContainer = styled.div({
    cursor: 'pointer',

    '& .mm-icon': {
        pointerEvents: 'none',
    },
});
