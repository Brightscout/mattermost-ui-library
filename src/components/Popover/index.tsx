import React from 'react';
import {OverlayTrigger, Popover as BootstrapPopover} from 'react-bootstrap';

type PopoverProps = {
    placement?: PopoverPlacement;
    popoverBody: JSX.Element;
    children: JSX.Element;
    className?: string;
}

const Popover = ({
    placement = 'bottom',
    children,
    popoverBody,
    className = '',
}: PopoverProps): JSX.Element => (
    <OverlayTrigger
        placement={placement}
        trigger='focus'
        delay={300}
        overlay={
            <BootstrapPopover
                id='popover'
                className={className}
            >
                {popoverBody}
            </BootstrapPopover>
        }
    >
        {children}
    </OverlayTrigger>
);

export default Popover;
