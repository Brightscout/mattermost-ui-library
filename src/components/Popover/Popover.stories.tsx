/* eslint-disable no-console */
import React from 'react';
import {Story, Meta} from '@storybook/react';

import {Icon} from '@Components/Icon';
import {MenuItem} from '@Components/MenuItem';

import {Popover} from './Popover.component';
import {PopoverProps} from './Popover';

export default {
    title: 'Component/Popover',
    component: Popover,
    parameters: {
        layout: 'centered',
    },
    args: {
        children: <Icon name='MenuDots'/>,
        placement: 'bottom-end',
        popoverBody: <>
            <MenuItem
                onClick={() => ''}
                leadingIcon='Edit'
                label='option 1'
            />
            <MenuItem
                onClick={() => ''}
                leadingIcon='Delete'
                label='option 2'
            />
        </>,
    },
} as Meta<PopoverProps>;

const PopoverTemplate:Story<PopoverProps> = (args) => <Popover {...args}/>;

// Default
export const Default = PopoverTemplate.bind({});
