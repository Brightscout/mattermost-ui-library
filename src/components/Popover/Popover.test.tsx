import React from 'react';
import {ReactWrapper, mount} from 'enzyme';

import Popover from '.';

// Mock props data
const PopoverChildrenClass = 'popover-children-class';
const PopoverBodyClass = 'popover-body-class';
const PopoverChildren = <span className={PopoverChildrenClass}>{'Mock Svg'}</span>;
const PopoverClassName = 'popover-class';
const PopoverBody = <span className={PopoverBodyClass}>{'Popover body'}</span>;

describe('Popover', () => {
    let component: ReactWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

    beforeEach(() => {
        component = mount(
            <Popover
                className={PopoverClassName}
                popoverBody={PopoverBody}
            >
                {PopoverChildren}
            </Popover>,
        );
    });

    afterEach(() => {
        component.unmount();
    });

    it('Should render correctly', () => {
        expect(component).toMatchSnapshot();
    });

    it('Should render the children correctly', () => {
        expect(component.find(`.${PopoverChildrenClass}`).exists()).toBeTruthy();
    });
});
