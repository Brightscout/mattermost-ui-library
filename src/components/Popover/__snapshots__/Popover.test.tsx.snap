// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Should render correctly and matches snapshot 1`] = `ReactWrapper {}`;
